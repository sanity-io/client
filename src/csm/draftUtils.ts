// nominal/opaque type hack
type Opaque<T, K> = T & {__opaqueId__: K}

/** @internal */
export type DraftId = Opaque<string, 'draftId'>

/** @internal */
export type PublishedId = Opaque<string, 'publishedId'>

/** @internal */
export const DRAFTS_FOLDER = 'drafts'

/** @internal */
export const VERSION_FOLDER = 'versions'

const PATH_SEPARATOR = '.'
const DRAFTS_PREFIX = `${DRAFTS_FOLDER}${PATH_SEPARATOR}`
const VERSION_PREFIX = `${VERSION_FOLDER}${PATH_SEPARATOR}`

/** @internal */
export function isDraftId(id: string): id is DraftId {
  return id.startsWith(DRAFTS_PREFIX)
}

/** @internal */
export function isVersionId(id: string): boolean {
  return id.startsWith(VERSION_PREFIX)
}

/** @internal */
export function isPublishedId(id: string): id is PublishedId {
  return !isDraftId(id) && !isVersionId(id)
}

/** @internal */
export function getDraftId(id: string): DraftId {
  if (isVersionId(id)) {
    const publishedId = getPublishedId(id)
    return (DRAFTS_PREFIX + publishedId) as DraftId
  }

  return isDraftId(id) ? id : ((DRAFTS_PREFIX + id) as DraftId)
}

/**  @internal */
export function getVersionId(id: string, version: string): string {
  if (version === 'drafts' || version === 'published') {
    throw new Error('Version can not be "published" or "drafts"')
  }

  return `${VERSION_PREFIX}${version}${PATH_SEPARATOR}${getPublishedId(id)}`
}

/**
 *  @internal
 *  Given an id, returns the versionId if it exists.
 *  e.g. `versions.summer-drop.foo` = `summer-drop`
 *  e.g. `drafts.foo` = `undefined`
 *  e.g. `foo` = `undefined`
 */
export function getVersionFromId(id: string): string | undefined {
  if (!isVersionId(id)) return undefined
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [_versionPrefix, versionId, ..._publishedId] = id.split(PATH_SEPARATOR)

  return versionId
}

/** @internal */
export function getPublishedId(id: string): PublishedId {
  if (isVersionId(id)) {
    // make sure to only remove the versions prefix and the bundle name
    return id.split(PATH_SEPARATOR).slice(2).join(PATH_SEPARATOR) as PublishedId as PublishedId
  }

  if (isDraftId(id)) {
    return id.slice(DRAFTS_PREFIX.length) as PublishedId
  }

  return id as PublishedId
}
