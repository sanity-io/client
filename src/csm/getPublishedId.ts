export const DRAFTS_PREFIX = 'drafts.'

/** @internal */
export function getPublishedId(id: string): string {
  if (id.startsWith(DRAFTS_PREFIX)) {
    return id.slice(DRAFTS_PREFIX.length)
  }

  return id
}
