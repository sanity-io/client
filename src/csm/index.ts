import * as studioPath from './studioPath'
export {studioPath}
export * from './applySourceDocuments'
export {createEditUrl} from './createEditUrl'
export {
  type DraftId,
  DRAFTS_FOLDER,
  getDraftId,
  getPublishedId,
  getVersionFromId,
  getVersionId,
  isDraftId,
  isPublishedId,
  isVersionId,
  type PublishedId,
  VERSION_FOLDER,
} from './draftUtils'
export {jsonPath, jsonPathToStudioPath, parseJsonPath, studioPathToJsonPath} from './jsonPath'
export {resolvedKeyedSourcePath} from './resolvedKeyedSourcePath'
export {resolveEditInfo} from './resolveEditInfo'
export {resolveEditUrl} from './resolveEditUrl'
export {resolveMapping} from './resolveMapping'
export type * from './types'
export {walkMap} from './walkMap'
