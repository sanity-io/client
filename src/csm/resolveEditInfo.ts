import {parseJsonPath} from './jsonPath'
import {resolveMapping} from './resolveMapping'
import type {
  CreateEditUrlOptions,
  ResolveEditInfoOptions,
  StudioBaseRoute,
  StudioBaseUrl,
  StudioUrl,
} from './types'

/** @internal */
export function resolveEditInfo(options: ResolveEditInfoOptions): CreateEditUrlOptions | undefined {
  const {resultSourceMap: csm, resultPath} = options
  const {mapping, pathSuffix} = resolveMapping(resultPath, csm) || {}

  if (!mapping) {
    // console.warn('no mapping for path', { path: resultPath, sourceMap: csm })
    return undefined
  }

  if (mapping.source.type === 'literal') {
    return undefined
  }

  if (mapping.source.type === 'unknown') {
    return undefined
  }

  const sourceDoc = csm.documents[mapping.source.document]
  const sourcePath = csm.paths[mapping.source.path]

  if (sourceDoc && sourcePath) {
    const {baseUrl, workspace, tool} = resolveStudioBaseRoute(
      typeof options.studioUrl === 'function' ? options.studioUrl(sourceDoc) : options.studioUrl,
    )
    if (!baseUrl) return undefined
    const {_id, _type} = sourceDoc
    return {
      baseUrl,
      workspace,
      tool,
      id: _id,
      type: _type,
      path: parseJsonPath(sourcePath + pathSuffix),
    } satisfies CreateEditUrlOptions
  }

  return undefined
}

/** @internal */
export function resolveStudioBaseRoute(studioUrl: StudioUrl): StudioBaseRoute {
  let baseUrl: StudioBaseUrl = typeof studioUrl === 'string' ? studioUrl : studioUrl.baseUrl
  if (baseUrl !== '/') {
    baseUrl = baseUrl.replace(/\/$/, '')
  }
  if (typeof studioUrl === 'string') {
    return {baseUrl}
  }
  return {...studioUrl, baseUrl}
}
