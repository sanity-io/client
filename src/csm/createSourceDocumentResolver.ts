import {getDraftId, getPublishedId, getVersionId} from './draftUtils'
import {resolvePerspectives} from './resolvePerspectives'
import type {ClientPerspective, ContentSourceMapDocuments, SanityDocument} from './types'

/** @internal */
export type ResolvedDocument = Partial<SanityDocument> &
  Required<Pick<SanityDocument, '_id' | '_type'>>

/** @internal */
export type MatchedDocument = Partial<SanityDocument> &
  Required<Pick<SanityDocument, '_id' | '_type' | '_originalId'>>

/** @internal */
export function createSourceDocumentResolver(
  getCachedDocument: (
    sourceDocument: ContentSourceMapDocuments[number],
  ) => ResolvedDocument | null | undefined,
  _perspective: Exclude<ClientPerspective, 'raw'>,
) {
  const perspectives = resolvePerspectives(_perspective)
  function findDocument(sourceDocument: ContentSourceMapDocuments[number]) {
    for (const perspective of perspectives) {
      let match: ReturnType<typeof getCachedDocument> = null
      if (perspective.startsWith('r')) {
        match = getCachedDocument({
          ...sourceDocument,
          _id: getVersionId(sourceDocument._id, perspective),
        })
      }
      if (perspective === 'drafts') {
        match = getCachedDocument({
          ...sourceDocument,
          _id: getDraftId(sourceDocument._id),
        })
      }
      if (perspective === 'published') {
        match = getCachedDocument({
          ...sourceDocument,
          _id: getPublishedId(sourceDocument._id),
        })
      }
      if (match) {
        return {...match, _id: getPublishedId(match._id), _originalId: match._id}
      }
    }
    return null
  }
  // define resolver that loops over source documents and perspectives
  return function resolveSourceDocument(
    sourceDocument: ContentSourceMapDocuments[number],
  ): MatchedDocument | null {
    return findDocument(sourceDocument)
  }
}
