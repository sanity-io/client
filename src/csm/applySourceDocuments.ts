import {DRAFTS_PREFIX, getPublishedId} from './getPublishedId'
import {parseJsonPath} from './jsonPath'
import {resolveMapping} from './resolveMapping'
import * as paths from './studioPath'
import type {
  Any,
  ApplySourceDocumentsUpdateFunction,
  ClientPerspective,
  ContentSourceMap,
  ContentSourceMapDocuments,
  Path,
  SanityDocument,
} from './types'
import {walkMap} from './walkMap'

const defaultUpdateFunction = <T = unknown>(changed: T): T => changed

/**
 * Optimistically applies source documents to a result, using the content source map to trace fields.
 * Can be used to apply mutations to documents being edited in a Studio, or any mutation on Content Lake, to a result with extremely low latency.
 * @alpha
 */
export function applySourceDocuments<Result = unknown>(
  result: Result,
  resultSourceMap: ContentSourceMap | undefined,
  getCachedDocument: (
    sourceDocument: ContentSourceMapDocuments[number],
  ) => Partial<SanityDocument> | null | undefined,
  updateFn: ApplySourceDocumentsUpdateFunction = defaultUpdateFunction,
  perspective: ClientPerspective = 'raw',
): Result {
  if (!resultSourceMap) return result

  if (perspective !== 'published' && perspective !== 'raw' && perspective !== 'previewDrafts') {
    throw new Error(`Unknown perspective "${perspective}"`)
  }

  return walkMap(JSON.parse(JSON.stringify(result)), (value, path) => {
    const resolveMappingResult = resolveMapping(path, resultSourceMap)
    if (!resolveMappingResult) {
      // console.warn('no mapping for path', path)
      return value
    }

    const {mapping, pathSuffix} = resolveMappingResult
    if (mapping.type !== 'value') {
      return value
    }

    if (mapping.source.type !== 'documentValue') {
      return value
    }

    const sourceDocument = resultSourceMap.documents[mapping.source.document]
    const sourcePath = resultSourceMap.paths[mapping.source.path]

    if (sourceDocument) {
      const parsedPath = parseJsonPath(sourcePath + pathSuffix)
      const stringifiedPath = paths.toString(parsedPath as Path)

      // The _id is sometimes used used as `key` in lists, and should not be changed optimistically
      if (stringifiedPath === '_id') {
        return value
      }

      let cachedDocument: Partial<SanityDocument> | null | undefined
      if (perspective === 'previewDrafts') {
        cachedDocument = getCachedDocument(
          sourceDocument._id.startsWith(DRAFTS_PREFIX)
            ? sourceDocument
            : {...sourceDocument, _id: `${DRAFTS_PREFIX}.${sourceDocument._id}}`},
        )
        if (!cachedDocument) {
          cachedDocument = getCachedDocument(
            sourceDocument._id.startsWith(DRAFTS_PREFIX)
              ? {...sourceDocument, _id: getPublishedId(sourceDocument._id)}
              : sourceDocument,
          )
        }
        if (cachedDocument) {
          cachedDocument = {
            ...cachedDocument,
            _id: getPublishedId(sourceDocument._id),
            _originalId: sourceDocument._id,
          }
        }
      } else {
        cachedDocument = getCachedDocument(sourceDocument)
      }

      if (!cachedDocument) {
        return value
      }

      const changedValue = cachedDocument
        ? paths.get<Result[keyof Result]>(cachedDocument, stringifiedPath, value)
        : value
      return value === changedValue
        ? value
        : updateFn<Result[keyof Result]>(changedValue as Any, {
            cachedDocument,
            previousValue: value as Result[keyof Result],
            sourceDocument,
            sourcePath: parsedPath,
          })
    }

    return value
  }) as Result
}
