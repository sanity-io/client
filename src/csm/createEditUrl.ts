import {DRAFTS_PREFIX, getPublishedId} from './getPublishedId'
import {jsonPathToStudioPath} from './jsonPath'
import * as studioPath from './studioPath'
import type {CreateEditUrlOptions, EditIntentUrl, StudioBaseUrl} from './types'

/** @internal */
export function createEditUrl(options: CreateEditUrlOptions): `${StudioBaseUrl}${EditIntentUrl}` {
  const {
    baseUrl,
    workspace: _workspace = 'default',
    tool: _tool = 'default',
    id: _id,
    type,
    path,
    projectId,
    dataset,
  } = options

  if (!baseUrl) {
    throw new Error('baseUrl is required')
  }
  if (!path) {
    throw new Error('path is required')
  }
  if (!_id) {
    throw new Error('id is required')
  }
  if (baseUrl !== '/' && baseUrl.endsWith('/')) {
    throw new Error('baseUrl must not end with a slash')
  }

  const workspace = _workspace === 'default' ? undefined : _workspace
  const tool = _tool === 'default' ? undefined : _tool
  const id = getPublishedId(_id)
  const stringifiedPath = Array.isArray(path)
    ? studioPath.toString(jsonPathToStudioPath(path))
    : path

  // eslint-disable-next-line no-warning-comments
  // @TODO Using searchParams as a temporary workaround until `@sanity/overlays` can decode state from the path reliably
  const searchParams = new URLSearchParams({
    baseUrl,
    id,
    type,
    path: stringifiedPath,
  })
  if (workspace) {
    searchParams.set('workspace', workspace)
  }
  if (tool) {
    searchParams.set('tool', tool)
  }
  if (projectId) {
    searchParams.set('projectId', projectId)
  }
  if (dataset) {
    searchParams.set('dataset', dataset)
  }
  if (_id.startsWith(DRAFTS_PREFIX)) {
    searchParams.set('isDraft', '')
  }

  const segments = [baseUrl === '/' ? '' : baseUrl]
  if (workspace) {
    segments.push(workspace)
  }
  const routerParams = [
    'mode=presentation',
    `id=${id}`,
    `type=${type}`,
    `path=${encodeURIComponent(stringifiedPath)}`,
  ]
  if (tool) {
    routerParams.push(`tool=${tool}`)
  }
  segments.push('intent', 'edit', `${routerParams.join(';')}?${searchParams}`)
  return segments.join('/') as unknown as `${StudioBaseUrl}${EditIntentUrl}`
}
