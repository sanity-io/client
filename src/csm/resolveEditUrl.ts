import {createEditUrl} from './createEditUrl'
import {studioPathToJsonPath} from './jsonPath'
import {resolveEditInfo} from './resolveEditInfo'
import type {ResolveEditUrlOptions} from './types'

/** @alpha */
export function resolveEditUrl(
  options: ResolveEditUrlOptions,
): ReturnType<typeof createEditUrl> | undefined {
  const {resultSourceMap, studioUrl} = options
  const resultPath = studioPathToJsonPath(options.resultPath)

  const editInfo = resolveEditInfo({
    resultPath,
    resultSourceMap,
    studioUrl,
  })
  if (!editInfo) {
    return undefined
  }

  return createEditUrl(editInfo)
}
