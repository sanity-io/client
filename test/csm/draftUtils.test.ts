import {describe, expect, it, test} from 'vitest'

import {getPublishedId, getVersionFromId, getVersionId} from '../../src/csm/draftUtils'

test.each([
  ['From published id', 'agot', 'summer-drop', 'versions.summer-drop.agot'],
  ['From draft id', 'drafts.agot', 'summer-drop', 'versions.summer-drop.agot'],
  ['From same version id', 'versions.summer-drop.agot', 'summer-drop', 'versions.summer-drop.agot'],
  [
    'From other version id',
    'versions.winter-drop.agot',
    'summer-drop',
    'versions.summer-drop.agot',
  ],
])('getVersionId(): %s', (_, documentId, equalsDocumentId, shouldEqual) => {
  expect(getVersionId(documentId, equalsDocumentId)).toEqual(shouldEqual)
})

test.each([
  ['from published id', 'agot', 'agot'],
  ['from draft id', 'drafts.agot', 'agot'],
  ['from version id', 'versions.summer-drop.agot', 'agot'],
  ['from complex id with version', 'versions.summer-drop.foo.agot', 'foo.agot'],
])('getPublishedId(): %s', (_, documentId, shouldEqual) => {
  expect(getPublishedId(documentId)).toEqual(shouldEqual)
})

describe('getVersionFromId', () => {
  it('should return the bundle slug', () => {
    expect(getVersionFromId('versions.summer.my-document-id')).toBe('summer')
  })

  it('should return the undefined if no bundle slug is found and document is a draft', () => {
    expect(getVersionFromId('drafts.my-document-id')).toBe(undefined)
  })

  it('should return the undefined if no bundle slug is found and document is published', () => {
    expect(getVersionFromId('my-document-id')).toBe(undefined)
  })
})
