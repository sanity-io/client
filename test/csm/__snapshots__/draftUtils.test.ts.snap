// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getVersionId > getVersionId(drafts) throws 1`] = `[Error: Version can not be "published" or "drafts"]`;

exports[`getVersionId > getVersionId(published) throws 1`] = `[Error: Version can not be "published" or "drafts"]`;
